### PR TITLE
feat: update charmcraft to use craft-application from main

### PR DESCRIPTION
**Pull request by lengau**

### CODE REVIEW

**Comment by syu-w** [comment]
craft-application 2.0 already?

File: charmcraft/parts/charm.py [hunk 1]
```diff
@@ -265,11 +265,16 @@ def get_build_packages(self) -> set[str]:
 
     def get_build_environment(self) -> dict[str, str]:
         """Return a dictionary with the environment to use in the build step."""
+        environment = {
+            # Cryptography fails to load OpenSSL legacy provider in some circumstances.
+            # Since we don't need the legacy provider, this works around that bug.
```
Comment (mr-cal) [inline charmcraft/parts/charm.py:270]: If you have a bug to link to, it would be nice to add here

**Review by mr-cal — APPROVED** [review]
LGTM

File: charmcraft/parts/charm.py (hunk 1 above)
Comment (lengau) [reply charmcraft/parts/charm.py:270]: Sadly not. It's something that only appears once you snap the app (so it's possibly patchelf related). Problem is snapcraft uses a very old version of patchelf. I suppose I could make an issue on snapcraft about it?

**Comment by lengau** [comment]
> craft-application 2.0 already?

Yeah this is for the `main` branch @syu-w

**Pull request merged**